apps/ejabberd/src/mod_mam_utils.erl

```diff
@@ -668,6 +680,34 @@ form_field(Type, VarName) ->
                     {<<"var">>, VarName}]}.
 
 %% -----------------------------------------------------------------------
+%% Text search tokenization
+
+%% -----------------------------------------------------------------------
+%% @doc
+%% Normalize given text to improve text search in some MAM backends.
+%% This normalization involves making text all lowercase, replacing some word separators
+%% ([, .:;-?!]) with given one (by default "%") and removing all unicode characters that are
+%% considered non-alphanumerical.
+%% For example, text: "My cat, was eaten by: my dog?!? Why...?!?" will be normalized as:
+%% "my%cat%was%eaten%by%my%dog%why"
+%% @end
+%% -----------------------------------------------------------------------
+-spec normalize_search_text(binary() | string() | undefined) -> string() | undefined.
+normalize_search_text(Text) ->
+    normalize_search_text(Text, "%").
+
+-spec normalize_search_text(binary() | string() | undefined, string()) -> string() | undefined.
+normalize_search_text(undefined, _WordSeparator) ->
+    undefined;
+normalize_search_text(Text, WordSeparator) ->
```

Comment:
This function has to do some heavy lifting on texts, have you measured the performance impact of <full-text-search on>?